File: src/Variables.java
import java.util.Map;

/**
 * Created by pranavkonduru on 9/2/18.
 */
public class Variables {

    public static final Integer RANDOM_CONSTANT = 15132;
    public static final String RANDOM_STRING = "hello!";

    private Map<String, String> randomMap;

    /**
     * Constructor
     * @param randomMap
     */
    public Variables(Map<String, String> randomMap) {
        this.randomMap = randomMap;
    }

    private boolean isEmpty = false;
    /**
     * Custom getter with nullability checks
     * @return
     */
    public boolean isMapEmpty() {
        if(randomMap != null){
            isEmpty = randomMap.size() == 0;
        }
        return isEmpty;
    }


    private int counter;
    /**
     * Custom setter
     * @param counter
     */
    public void setCounter(int counter) {
        if(counter > 0) {
            this.counter = counter;
        }
    }
}

File: src/Variables.kt
/**
 * Created by pranavkonduru on 9/2/18.
 */

/**
 * Constructor is defined in the class definition
 * ? after type indicates that the variable can be null
 */
class Variables (var randomMap : Map<String, String>?) {

    /**
     * Custom getter.
     * If randomMap was null, it will not execute .size
     */
    val isMapEmpty get() = randomMap?.size == 1

    /**
     * Custom getter.
     * Backing field is needed to set fields
     */
    var counter = 0
        set(value) {
           if(value > 0) field = value
        }

    /**
     * Static methods or fields are defined
     */
    companion object {
        /**
         * Not a compile time constant.
         * Will generate getters and setters
         */
        val RANDOM_CONSTANT = 15132

        /**
         * Compile time constant
         */
        const val RANDOM_STRING = "hello!"
    }

}